WalletWasabi/BranchNBound/BranchAndBound.cs

```diff
@@ -0,0 +1,190 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.TransactionOutputs;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BranchNBound
+{
+	public class BranchAndBound
+	{
+		private readonly Random _random = new();
+
+		public BranchAndBound(List<SmartCoin> availableCoins)
+		{
+			Count = availableCoins.Count;
+			SortedUTXOs = availableCoins.OrderByDescending(x => x.Amount).ToArray();
+		}
+
+		private enum NextAction
+		{
+			/// <summary>First try to include a coin and then try not to include the coin in the selection.</summary>
+			AandB,
+
+			/// <summary>First try NOT to include a coin and then try to include the coin in the selection.</summary>
+			BandA,
+
+			/// <summary>Include coin.</summary>
+			A,
+
+			/// <summary>Omit coin.</summary>
+			B,
+
+			/// <summary>Current selection is wrong, rolling back and trying different combination.</summary>
+			Backtrack
+		}
+
+		/// <remarks>Sorted in desceding order.</remarks>
+		private SmartCoin[] SortedUTXOs { get; }
+
+		private int Count { get; }
+
+		public bool TryGetExactMatch(Money target, [NotNullWhen(true)] out List<SmartCoin>? selectedCoins)
+		{
+			using (BenchmarkLogger.Measure())
+			{
+				if (CalculateSum(SortedUTXOs) < target)
```

Comment:
Any reason not use simple:

```suggestion
				if (SortedUTXOs.Select(c => c.Amount).Sum() < target)
```

?

Given that `SortedUTXOs.Select(c => c.Amount).Sum()` is a constant really, it can be in the constructor.